datumaro/components/project.py

```diff
@@ -2,486 +2,2481 @@
+        if not dvcignore:
+            dvcignore = '.dvcignore'
+        repo_root = self._project_dir
+        dvcignore = osp.abspath(osp.join(repo_root, dvcignore))
+        assert dvcignore.startswith(repo_root), dvcignore
+
+        _update_ignore_file(paths, repo_root=repo_root,
+            mode=mode, filepath=dvcignore)
+
+    # This ruamel parser is needed to preserve comments,
+    # order and form (if multiple forms allowed by the standard)
+    # of the entries in the file. It can be reused.
+    yaml_parser = yaml.YAML(typ='rt')
+
+    @classmethod
+    def get_hash_from_dvcfile(cls, path) -> str:
+        with open(path) as f:
+            contents = cls.yaml_parser.load(f)
+        return contents['outs'][0]['md5']
+
+    FILE_HASH_LEN = 32
+    DIR_HASH_SUFFIX = '.dir'
+    DIR_HASH_LEN = FILE_HASH_LEN + len(DIR_HASH_SUFFIX)
+
+    @classmethod
+    def is_file_hash(cls, s: str) -> bool:
+        return len(s) == cls.FILE_HASH_LEN
+
+    @classmethod
+    def is_dir_hash(cls, s: str) -> bool:
+        return len(s) == cls.DIR_HASH_LEN and s.endswith(cls.DIR_HASH_SUFFIX)
+
+    @classmethod
+    def is_hash(cls, s: str) -> bool:
+        return cls.is_file_hash(s) or cls.is_dir_hash(s)
+
+    def write_obj(self, obj_hash, dst_dir, allow_links=True):
+        def _copy_obj(src, dst, link=False):
+            os.makedirs(osp.dirname(dst), exist_ok=True)
+            if link:
+                os.link(src, dst)
+            else:
+                shutil.copy(src, dst, follow_symlinks=True)
+
+        src = self.obj_path(obj_hash)
+        if osp.isfile(src):
+            _copy_obj(src, dst_dir, link=allow_links)
+            return
+
+        src += self.DIR_HASH_SUFFIX
+        if not osp.isfile(src):
+            raise UnknownRefError(obj_hash)
+
+        with open(src) as f:
+            src_meta = json.load(f)
+        for entry in src_meta:
+            _copy_obj(self.obj_path(entry['md5']),
+                osp.join(dst_dir, entry['relpath']), link=allow_links)
+
+    def remove_cache_obj(self, obj_hash: str):
+        src = self.obj_path(obj_hash)
+        if osp.isfile(src):
+            os.remove(src)
+            return
+
+        src += self.DIR_HASH_SUFFIX
+        if not osp.isfile(src):
+            raise UnknownRefError(obj_hash)
+
+        with open(src) as f:
+            src_meta = json.load(f)
+        for entry in src_meta:
+            entry_path = self.obj_path(entry['md5'])
+            if osp.isfile(entry_path):
+                os.remove(entry_path)
+
+        os.remove(src)
+
+class Tree:
+    # can be:
+    # - attached to the work dir
+    # - attached to a revision
+
+    def __init__(self, project: 'Project',
+            config: Optional[TreeConfig] = None,
+            rev: Optional[str] = None):
+        assert isinstance(project, Project)
+        assert not rev or project.is_ref(rev), rev
+
+        if not isinstance(config, TreeConfig):
+            config = TreeConfig(config)
```

Comment:
According to the type annotations, the only thing `config` can be other than a `TreeConfig` is `None`, and while it does seem like this code will work for `None`, it would be more transparent if written as:

```python
if config is None:
    config = TreeConfig()
```

Or perhaps even:

```python
config = config or TreeConfig()
```

Reply:
Any place where a Config subclass is expected, can also accept Config and dict. Count it a project convention. Changed the type annotations.